wagtail/wagtailadmin/navigation.py

```diff
@@ -0,0 +1,103 @@
+    if user.is_superuser:
+        # superuser has implicit permission on the root node
+        pages_with_direct_permission = Page.objects.filter(depth=1)
+    else:
+        pages_with_direct_permission = Page.objects.filter(
+            group_permissions__group__in=user.groups.all(),
+            group_permissions__permission_type__in=['add', 'edit', 'publish', 'lock']
+        )
+
+    if not(pages_with_direct_permission):
+        return []
+
+    # Find the closest common ancestor of the pages the user has permission for;
+    # this (or its children) will be the root menu level for this user.
+    cca_path = pages_with_direct_permission[0].path
+    for page in pages_with_direct_permission[1:]:
+        # repeatedly try shorter prefixes of page.path until we find one that cca_path starts with;
+        # this becomes the new cca_path
+        for path_len in range(len(page.path), 0, -Page.steplen):
+            path_to_test = page.path[0:path_len]
+            if cca_path.startswith(path_to_test):
+                cca_path = path_to_test
+                break
+
+    # Determine the depth (within the overall page tree) at which this user's menu starts:
+    # * if CCA is the root node, start at depth 2 (immediate children of root - because we
+    #   never want to show root in the navigation);
+    # * else if CCA is a node they have direct permission for, start at that depth
+    #   (so that they can edit that root node)
+    # * else start one level deeper (because the root node is only needed to provide navigation
+    #   to deeper levels, and one level deeper is the first point where there's a choice to make)
+
+    if len(cca_path) == Page.steplen:
+        # CCA is the root node
+        menu_root_depth = 2
+    elif any(page.path == cca_path for page in pages_with_direct_permission):
+        # user has direct permission on the CCA node
+        menu_root_depth = int(len(cca_path) / Page.steplen)
+    else:
+        menu_root_depth = int(len(cca_path) / Page.steplen) + 1
+
+    # Run the query to fetch all pages to be shown in the navigation. This consists of the following
+    # set of pages, for each page in pages_with_direct_permission:
+    #  * all ancestors (plus self) from menu_root_depth down
+    #  * all descendants that have children
+    #  * all descendants at menu_root_depth, regardless of whether they have children.
+    #    (this ensures that a freshly built site with no child pages won't result in an empty menu)
+
+    ancestor_paths = [
+        page.path[0:path_len]
+        for page in pages_with_direct_permission
+        for path_len in range(menu_root_depth * Page.steplen, len(page.path) + Page.steplen, Page.steplen)
+    ]
+
+    criteria = Q(path__in=ancestor_paths)
+
+    for page in pages_with_direct_permission:
+        criteria = criteria | (
+            Q(path__startswith=page.path) & (
+                Q(depth=menu_root_depth) | Q(numchild__gt=0)
```

Comment:
OK - the `Q(depth=menu_root_depth)` clause can only ever take effect in the specific case where the user has permission on the root, and the menu starts one level down from that (i.e. the special case at https://github.com/torchbox/wagtail/pull/2869/files#diff-c1b04653cef1c28e847039a46736aca8R44)

It can therefore be simplified to `Q(depth=2)`, which sure enough is exactly what the current code does :-) https://github.com/torchbox/wagtail/pull/2869/files#diff-9ec0dccfdaa33a0fd310c76b5db2cb37L1350

This clause is important because it's the only thing that ensures that a single-page site at the top level shows in the menu - without it, a superuser on a freshly built Wagtail instance won't get a menu at all. (This is covered by the unit tests at https://github.com/torchbox/wagtail/pull/2869/files#diff-f8ac0fd58cb50cefaa757703fe34c15fR61 - page 10, `/home-2/`, is a single-page site.) I thought this needed to be generalised to deal with the case of a non-superuser who has permission for a single leaf node somewhere deeper in the tree - but actually it doesn't, because `ancestor_paths` is inclusive (and therefore the node in question is counted as an ancestor of itself).